lib/internal/fs/dir.js

```diff
@@ -0,0 +1,162 @@
+const internalUtil = require('internal/util');
+const {
+  getDirent,
+  getOptions,
+  getValidatedPath,
+  handleErrorFromBinding,
+  makeCallback
+} = require('internal/fs/utils');
+
+const kDirHandle = Symbol('Dir handle');
+const kDirPath = Symbol('Dir path');
+const kDirOptions = Symbol('Dir options');
+const kDirReadP = Symbol('Dir Awaitable Read');
+const kDirCloseP = Symbol('Dir Awaitable Close');
+
+class Dir {
+  constructor(handle, path, options) {
+    if (handle == null) throw new ERR_MISSING_ARGS('handle');
+    this[kDirHandle] = handle;
+    this[kDirPath] = path;
+
+    this[kDirOptions] = getOptions(options, {
+      encoding: 'utf8'
+    });
+
+    this[kDirReadP] = internalUtil.promisify(this.read.bind(this));
+    this[kDirCloseP] = internalUtil.promisify(this.close.bind(this));
+  }
+
+  get path() {
+    return pathModule.toNamespacedPath(this[kDirPath]);
+  }
+
+  read(options, callback) {
+    callback = typeof options === 'function' ? options : callback;
+    if (!callback) {
+      return this[kDirReadP](options);
+    }
+    callback = makeCallback(callback);
+    options = getOptions(options, this[kDirOptions]);
+
+    const req = new FSReqCallback();
+    req.oncomplete = (err, result) => {
+      if (err || result === null) {
+        return callback(err, result);
+      }
+      getDirent(this[kDirPath], result[0], result[1], callback);
+    };
+
+    this[kDirHandle].read(
+      options.encoding,
+      req
+    );
+  }
+
+  readSync(options) {
+    options = getOptions(options, this[kDirOptions]);
+
+    const ctx = { path: this[kDirPath] };
+    const result = this[kDirHandle].read(
+      options.encoding,
+      undefined,
+      ctx
+    );
+    handleErrorFromBinding(ctx);
+
+    if (result === null) {
+      return result;
+    }
+
+    return getDirent(this[kDirPath], result[0], result[1]);
+  }
+
+  close(callback) {
+    if (!callback) {
+      return this[kDirCloseP]();
+    }
+    callback = makeCallback(callback);
+    const req = new FSReqCallback();
+    req.oncomplete = callback;
+    this[kDirHandle].close(req);
+  }
+
+  closeSync() {
+    const ctx = { path: this[kDirPath] };
+    const result = this[kDirHandle].close(undefined, ctx);
+    handleErrorFromBinding(ctx);
+    return result;
+  }
+}
+
+Dir.prototype[Symbol.asyncIterator] = async function* dirReadAsyncIterGen() {
+  while (true) {
+    const result = await this[kDirReadP]();
+    if (result === null) {
+      this[kDirCloseP]();
+      return;
+    }
+    yield result;
+  }
+};
+
+function opendir(path, options, callback) {
+  callback = makeCallback(typeof options === 'function' ? options : callback);
+  path = getValidatedPath(path);
+  options = getOptions(options, {
+    encoding: 'utf8'
+  });
+
+  function opendirCallback(error, handle) {
+    if (error) {
+      callback(error);
+    } else {
+      callback(null, new Dir(handle, path, options));
+    }
+  }
+
+  const req = new FSReqCallback();
```

Comment:
… This needs to be associated with a `DirReq` instance on the C++ level, right? Otherwise we’re running into memory error trouble?